### DISCUSSION TITLE
how to set up business rules?

### DISCUSSION THREAD

**Discussion by 4rm1n** [pocketbase/pocketbase#962] [Q&A]
Hello,

I am quite lost on how to implement business rules when creating a new record.

Here is my specific case: I have three collections:

Users:
- Id
- Email
- Password

Races :
- Id
- Email
- Start date
- End date

Participation
- Id
- RaceId
- UserId
- ParticipationType

I want to prevent the creation of a new Participation if a User has already a Participation in a Race that takes place at the same time as another race, what is not clear is how I can handle values from other collections. Is there a specific syntax for this?

Thanks in advance

**Comment by ganigeorgiev** [comment]
Hm, I'm not sure if it can be expressed with just a simple API rule (or at least I can't think of how at the moment). Just for info, the documentation will be updated very soon with the v0.8.0 release, but you could find some information about the filter rules in https://pocketbase.io/docs/manage-collections/#collection-api-rules.

While the requirement may seems simple at first, in practice it is not so easy and even with raw sql you still have to do multiple joins or several selects to apply the validation criteria.

The problem is that while you can access related and non-related collections with dot-notation and `@collection.*`, currently there is no "all-joined-rows match" filter expression (or alternatively - a group exclusion rule). The following for example **will not work** because it will return true if there is at least one matching record, but what we want is every joined record from the "races" collection to match the filter conditions:

```
@request.data.userId = @request.auth.id &&
@request.data.userId = @collection.participation.userId &&
@request.data.raceId != @collection.participation.raceId &&
(@request.data.startDate < @collection.participation.raceId.startDate || @request.data.endDate > @collection..participation.raceId.endDate)
```

Therefore at the moment you'll have to either change your data structure/requirements OR use the [record before create/update Go hooks](https://pocketbase.io/docs/event-hooks/#record-hooks).

This is an example how the `OnRecordBeforeCreateRequest` hook could look like (_I haven't tested it_):
```go
app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
    if e.Record.Collection().Name != "participation" {
        return nil // not a participation collection
    }

    relatedRace, err := app.Dao().FindRecordById("races", e.Record.GetString("raceId"))
    if err != nil {
        return err
    }

    var totalOverlapping int
    countErr := app.Dao().DB().NewQuery(`
        select count(distinct participation.id)
        from participation
        left join races on races.id = participation.raceId
        where (
            races.id != {:raceId} AND
            participation.userId = {:userId} AND
            races.startDate >= {:startDate} AND
            races.endDate <= {:endDate}
        )
    `).Bind(dbx.Params{
        "raceId":    e.Record.GetString("raceId"),
        "userId":    e.Record.GetString("userId"),
        "startDate": relatedRace.GetString("startDate"),
        "endDate":   relatedRace.GetString("endDate"),
    }).Row(&totalOverlapping)

    if countErr != nil || totalOverlapping > 0 {
        return apis.NewBadRequestError("The user already participate in other races during the selected time period.", countErr)
    }

    return nil
})
```

_After v0.8.0, improving the filters and allowing more complex use cases will be one of main focus in the next couple releases (that include IN operator(s), multi-match expressions, accessing nested relation fields through `@request.data.*`, etc.)._
